pkg/services/ngalert/schedule/registry.go

```diff
@@ -111,3 +111,52 @@ type evaluation struct {
 	scheduledAt time.Time
 	version     int64
 }
+
+type schedulableAlertRulesRegistry struct {
+	rules map[models.AlertRuleKey]*models.SchedulableAlertRule
+	mu    sync.Mutex
+}
+
+// all returns all rules in the registry.
+func (r *schedulableAlertRulesRegistry) all() []*models.SchedulableAlertRule {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	result := make([]*models.SchedulableAlertRule, 0, len(r.rules))
+	for _, rule := range r.rules {
+		result = append(result, rule)
+	}
+	return result
+}
+
+func (r *schedulableAlertRulesRegistry) get(k models.AlertRuleKey) *models.SchedulableAlertRule {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.rules[k]
+}
+
+// set replaces all rules in the registry.
+func (r *schedulableAlertRulesRegistry) set(rules []*models.SchedulableAlertRule) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.rules = make(map[models.AlertRuleKey]*models.SchedulableAlertRule)
+	for _, rule := range rules {
+		r.rules[rule.GetKey()] = rule
+	}
+}
+
+// update inserts or replaces a rule in the registry.
+func (r *schedulableAlertRulesRegistry) update(rule *models.SchedulableAlertRule) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.rules[rule.GetKey()] = rule
+}
+
+func (r *schedulableAlertRulesRegistry) del(k models.AlertRuleKey) (*models.SchedulableAlertRule, bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	rule, ok := r.rules[k]
+	if ok {
+		delete(r.rules, k)
+	}
+	return rule, ok
+}
```

Comment:
```suggestion
func (r *schedulableAlertRulesRegistry) del(k models.AlertRuleKey) {
	r.mu.Lock()
	defer r.mu.Unlock()
	rule, ok := r.rules[k]
	if ok {
		delete(r.rules, k)
	}
}
```

Reply:
I think this implementation needs to be slightly different to the one in `alertRuleInfoRegistry`.

`alertRuleInfoRegistry` uses the delete operation to also stop the rule - we don't have to do that here. (I also think this opens the door to a potential refactor of using `get` and `del` for the stopping operation as opposed to just a `del` but there's another topic for another day and def not part of this PR)

Reply:
I like that both registries have similar function signatures. We don't need to remove the return values just because we don't use them?